### ISSUE TITLE
Elsa 2.0: Context (ContextId, ContextType)

### ISSUE THREAD

**Issue by Rafael94** [elsa-workflows/elsa-core#514]
Hi,

I work a lot with entities in my projects (tickets, devices, people, etc.)

Users can create their own workflows and therefore there can be multiple instances for one entity.

CorrelationId must exist only once ( `Task<WorkflowInstance?> GetByCorrelationIdAsync(string correlationId, WorkflowStatus status,CancellationToken cancellationToken = default)`). That's why I would like to work with the `ContextId`. But at the moment I can't specify a type (ticket, device, ...) for the instance. I could only help myself out by applying such a construct `ContextId = $"Ticket_{TicketId}"`.

My suggestion:
- In the WorkflowDefinition and in the WorkflowInstance store the optional `ContextType` (ticket, device). For the entity trigger, I introduced `EntityNameAttribute`. It could be moved to `Elsa.Core` to determine the `ContextType`.
- `Task<IEnumerable<WorkflowInstance>> ListByContextAsync(string contextType, string? contextId, string? tenantId,CancellationToken cancellationToken = default);` add to `IWorkflowInstanceStore`.

For later:
- Filter activities and triggers by `ContextType` so that only general and context specific activities and triggers are shown.

**Comment by sfmskywalker** [comment]
Your scenario makes sense, but I'm wondering if we could maybe  do something with custom "tags", allowing you to associate workflow instances with zero or more tags? The advantage of that is that other applications can associate more application-specific information with their workflows.

On the other hand, the `ContextId` is clearly intended for use with the subject/context of the workflow, so it might make sense to include its type as well. In fact, this would be a hard-requirement if we are to implement intellisense in expressions editing UIs.

**Comment by Rafael94** [comment]
Do you mean tags or properties? Tags is just a list of keywords. With properties you can store values in addition to the keywords.

Node: Node-1
SystemWorkflow: true
...

As you say, we already have `ContextId` for contextual workflows, so it would make sense to store the type as well. 

The UI is another issue. For now, I'll try to improve the backend based on your excellent preliminary work according to my wishes and share them.  As a step by step improve the backend and then the UI.

**Comment by Rafael94** [comment]
Can be closed, has been implemented.
